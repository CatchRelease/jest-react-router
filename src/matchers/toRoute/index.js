import { matcherHint, printExpected, printReceived } from 'jest-matcher-utils';

const toRoute = (router, path, { to }) => {
  const route = router
    .find('Route')
    .filterWhere(node => node.prop('path') === path);
  const expected = to.displayName || to.name;

  if (route.length === 0) {
    return {
      message: () => `No route matches ${path}`,
      pass: false
    };
  }

  if (route.props().component !== to) {
    const actual = route.props().component;
    const received = actual.displayName || actual.name;

    const message = `
      ${matcherHint('.toRoute')}

      Expected to route to:
        ${printExpected(expected)}
      Received:
        ${printReceived(received)}
    `;

    return {
      message: () => message,
      pass: false
    };
  }

  return {
    message: () => `Found ${path} routing to ${expected}`,
    pass: true
  };
};

export default toRoute;
