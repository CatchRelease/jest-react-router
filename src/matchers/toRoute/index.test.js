import React from 'react';
import { shallow } from 'enzyme';

import toRoute from '.';

expect.extend({ toRoute });

const Route = () => null;
const MyComponent = () => null;
const OtherComponent = () => null;

const router = shallow(
  <div>
    <Route path="/sign_in" component={MyComponent} />
  </div>
);

describe('.toRoute', () => {
  it('passes when a matching route is found', () => {
    expect(router).toRoute('/sign_in', { to: MyComponent });
  });

  it('fails when a route is not found at all', () => {
    expect(() =>
      expect(router).toRoute('/missing', { to: MyComponent })
    ).toThrowErrorMatchingSnapshot();
  });

  it('fails when a route is found with a different component', () => {
    expect(() =>
      expect(router).toRoute('/sign_in', { to: OtherComponent })
    ).toThrowErrorMatchingSnapshot();
  });
});

describe('.not.toRoute', () => {
  it('passes when a route is not found', () => {
    expect(router).not.toRoute('/missing', { to: MyComponent });
  });

  it('passes when a route exists for a different component', () => {
    expect(router).not.toRoute('/sign_in', { to: OtherComponent });
  });

  it('fails when a route is found', () => {
    expect(() =>
      expect(router).not.toRoute('/sign_in', { to: MyComponent })
    ).toThrowErrorMatchingSnapshot();
  });
});
