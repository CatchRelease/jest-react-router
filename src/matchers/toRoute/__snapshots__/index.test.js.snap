// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.not.toRoute fails when a route is found 1`] = `"Found /sign_in routing to MyComponent"`;

exports[`.toRoute fails when a route is found with a different component 1`] = `
"
      [2mexpect([22m[31mreceived[39m[2m).toRoute([22m[32mexpected[39m[2m)[22m

      Expected to route to:
        [32m\\"OtherComponent\\"[39m
      Received:
        [31m\\"MyComponent\\"[39m
    "
`;

exports[`.toRoute fails when a route is not found at all 1`] = `"No route matches /missing"`;
